File: playground/python/search/astar_test.py
import astar

class TestProblem:
    target = (253,252)
    def initial_state(self):
        return (0,0)
    def is_goal(self,state):
        return state == self.target
    def heuristic(self,state):
        (x,y) = state
        (tx,ty) = self.target 
        return (abs(x-tx)+4)/5 + (abs(y-ty)+2)/3
    def generate_moves(self,state):
        (x,y) = state
        yield (1,"up",(x,y+3))
        yield (1,"down",(x,y-2))
        yield (1,"left",(x-3,y))
        yield (1,"right",(x+5,y))

def test():
    sol = astar.path_to_list(astar.astar(TestProblem()).next())
    print len(sol)
    print sol

test()

File: playground/ocaml/search/AStarTest.ml
module Problem = struct
  type instance = unit
  type state = int * int
  type move = Left | Right | Up | Down
  module Cost = struct
    type t = int
    let compare (a,b) = a-b
    let add (a,b) = a+b
    let zero = 0
  end

  let goal = (243,242)
  let initial_state () = (0,0)
  let is_goal (_,state) = state = goal
  let move_cost (_,move) = 1
  let possible_moves (_,_) = [Up;Down;Left; Right]
  let heuristic (_,(x,y)) =
    let (tx,ty) = goal in (abs(x-tx)+4)/5 + (abs(y-ty)+2)/3
  let make_move (_,(x,y),move) =
    match move with
      | Left  -> (x-3,y)
      | Right -> (x+5,y)
      | Up    -> (x,y+3)
      | Down  -> (x,y-2)
end

module Solver = AStar.Make(Problem)

;;
let solver = Solver.make ()
;;
match Solver.next_solution solver with
    None -> print_string "No solution!\n"
  | Some (state,path,solver) ->
      print_string ("Found solution of length "^string_of_int (List.length path)^"\n")
